Contact journal day screen: New attributes (EXPOSUREAPP-5108)

d4rken opened the pull request:
Adds new attributes to person and location entry for each day.

Adjusting the style to 100% match the mockups might take some time, we could evaluate merging this earlier so other tickets tha depend on data entry can better be worked on. The functionality itself is already functional.

### TODO
- [x] A handful of tests (Extra PR)
- [x] Adjust style to fit mockups (Extra PR)

### Testing
* Compare with mockups
* Play with data entry, leaving the app, restoring it, updating the entries etc.

d4rken committed:
Strings for contact diary encounter/visit button groups and extensions.

d4rken committed:
First draft, enchance person and location items to allow changing the new attributes.

d4rken committed:
Merge branch 'feature/4741-contact-journal-extension-main' into feature/5108-contact-journal-day-screen-new-attributes

d4rken committed:
Move circumstances text input + button into a custom view.
Fix TextWatcher mechanism.
Reduce clickable container area to header.
Fix list items flickering on update.

d4rken committed:
Adjust duration storage to map long to JodaTime.Duration.
Improve list item update mechanism, we don't need live updates, just live restoration on stableId change.

LukasLechnerDev commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/day/tabs/common/DiaryCircumstancesTextView.kt:
```diff
@@ -0,0 +1,66 @@
+    private val infoButton: ImageView
+
+    private var afterTextChangedListener: ((String) -> Unit)? = null
+    private val circumStanceTextWatcher = object : TextWatcher {
```
Instead of `TextWatcher` with empty method implementations, you could use `TextView.doAfterTextChanged` from KTX: https://developer.android.com/kotlin/ktx/extensions-list#extension_functions_18

d4rken replied:
Using the focus approach for now, but thanks will use next time!

d4rken committed:
Fix inconsistent list/database updates.
We need to to refresh the list data consistently.
Introduce payload handling to prevent whole view refresh.

d4rken committed:
Call text changed listener after focus was lost.

d4rken committed:
LINTER ¯\_(ツ)_/¯

d4rken committed:
Fix instrumentation tests.

d4rken committed:
Trim text from circumstances input.

LukasLechnerDev commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/day/tabs/common/DiaryCircumstancesTextView.kt:
```diff
@@ -0,0 +1,58 @@
+                if (hasFocus) {
+                    Timber.v("Focused on %s", v)
+                } else {
+                    Timber.v("Lost focus on %s", v)
```
On my device the focus isn't lost when putting the app in background (and then close it) 🤔

d4rken replied:
Will offer a solution to this in a follow up PR.

d4rken committed:
Fix list items not expanding on selection because the ViewHolder wasn't re-layouted.

d4rken committed:
Merge branch 'feature/4741-contact-journal-extension-main' into feature/5108-contact-journal-day-screen-new-attributes

# Conflicts:
#	Corona-Warn-App/src/main/res/values/contact_diary_strings.xml

LukasLechnerDev reviewed (approved):
This was a triumph. 'Huge success'. It's hard to overstate my satisfaction.

axelherbstreith commented:
LGTM 👍  Tested on Samsung S8 and Huawei P30
